notebooks/sycamore-tutorial-intermediate-etl.ipynb

```diff
@@ -0,0 +1,410 @@
+   "metadata": {},
+   "source": [
+    "## Data formatting\n",
+    "\n",
+    "Sycamore enables you to apply user-defined functions over every document using the `map` and `map_batches` transforms. As an example, we might want to parse the date/time strings extracted earlier into proper date information, and we can use this code to do it:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from sycamore.data.document import Document\n",
+    "from dateutil import parser\n",
+    "def convert_timestamp(doc: Document) -> Document:\n",
+    "    try:\n",
+    "        if \"dateAndTime\" not in doc.properties['entity'] and \"dateTime\" not in doc.properties['entity']:\n",
+    "            return doc\n",
+    "        raw_date: str = doc.properties['entity'].get('dateAndTime') or doc.properties['entity'].get('dateTime')\n",
+    "        raw_date = raw_date.replace(\"Local\", \"\")\n",
+    "        parsed_date = parser.parse(raw_date, fuzzy=True)\n",
+    "        extracted_date = parsed_date.date()\n",
+    "        doc.properties['entity']['day'] = extracted_date.day\n",
+    "        doc.properties['entity']['month'] = extracted_date.month\n",
+    "        doc.properties['entity']['year'] = extracted_date.year\n",
+    "        if parsed_date.utcoffset():\n",
+    "            doc.properties['entity']['dateTime'] = parsed_date.isoformat()\n",
+    "        else:\n",
+    "            doc.properties['entity']['dateTime'] = parsed_date.isoformat() + \"Z\"\n",
+    "    except:\n",
+    "        pass\n",
+    "    return doc\n",
+    "\n",
+    "formatted_docset = chunked_docset.map(convert_timestamp)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Let's pause and look at what we've done and what's yet to do. So far, we've read a set of PDF documents from S3, partitioned them with the Aryn Partitioning Service, extracted some metadata from them, chunked them into different size elements, and performed some data cleaning.\n",
+    "\n",
+    "What's left is to embed our documents and write them to Pinecone. First, we'll copy the properties from each document to each element of that document using the `spread_properties` transform. Next, First, we will promote these elements to top-level document objects using the `explode` transform. We do this because Sycamore's `embed` transform only operates on documents, not elements."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "exploded_docset = formatted_docset.spread_properties([\"path\", \"entity\"]).explode()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now we'll embed each document with the `embed` transform, and generate term frequency tables for sparse vector search with the `term_frequency` transform. Pinecone requires this latter table."
```

Comment:
Pinecone requires the TF table if you want to do hybrid search in Pinecone. You don't need it tho (and tbh I'm not convinced that this actually gives you better relevance with the way pc implements - and forces us to implement - it)

Reply:
Ack

Reply:
Ack